cpp/src/decisiontree/batched-levelalgo/builder_base.cuh

```diff
@@ -27,10 +30,104 @@
 #include "split.cuh"
 
 #include <common/nvtx.hpp>
+#include <deque>
+#include <utility>
 
 namespace ML {
 namespace DT {
 
+template <typename NodeT>
+class NodeQueue {
+  const DecisionTreeParams params;
+  std::vector<NodeT> tree_;
+  std::deque<NodeWorkItem> work_items_;
+  int tree_depth_ = 0;
+  int num_leaves_ = 1;
+
+ public:
+  NodeQueue(DecisionTreeParams params, size_t max_nodes, size_t sampled_rows) : params(params)
+  {
+    tree_.reserve(max_nodes);
+    tree_.emplace_back(0, sampled_rows, 0);
+    if (!IsLeaf(tree_.back())) { work_items_.emplace_back(NodeWorkItem{0, 0, sampled_rows, 0}); }
+  }
+
+  int TreeDepth() { return tree_depth_; }
+  int NumLeaves() { return num_leaves_; }
+  std::vector<NodeT> GetTree() { return tree_; }
+
+  bool HasWork() { return work_items_.size() > 0; }
+
+  auto Pop()
+  {
+    std::vector<NodeWorkItem> result;
+    result.reserve(std::min(size_t(params.max_batch_size), work_items_.size()));
+    while (work_items_.size() > 0 && result.size() < params.max_batch_size) {
+      result.emplace_back(work_items_.front());
+      work_items_.pop_front();
+    }
+    return result;
+  }
+
+  bool IsLeaf(const NodeT& n)
+  {
+    if (n.depth >= params.max_depth) return true;
+    if (n.count < params.min_samples_split) return true;
+    if (params.max_leaves != -1 && num_leaves_ >= params.max_leaves) return true;
+    return false;
+  }
+
+  template <typename SplitT>
+  void Push(const std::vector<NodeWorkItem>& work_items, SplitT* h_splits)
+  {
+    // Update node queue based on splits
+    for (int i = 0; i < work_items.size(); i++) {
+      auto split      = h_splits[i];
+      auto item       = work_items[i];
+      const auto node = tree_[item.idx];
+      if (SplitNotValid(split, params.min_impurity_decrease, params.min_samples_leaf, node.count)) {
+        continue;
+      }
+
+      if (params.max_leaves != -1 && num_leaves_ >= params.max_leaves) break;
+
+      // parent
+      tree_[item.idx] = NodeT::CreateSplit(split.colid,
+                                           split.quesval,
+                                           split.best_metric_val,
+                                           tree_.size(),
+                                           node.count,
+                                           node.info.unique_id);
+      num_leaves_++;
```

Comment:
I am not sure why leaf count is incremented here. Is it utilizing the property of binary tree that number of intermediate and leaf nodes in a tree are same? If so, may be we can consider adding a line of comment to describe this clever idea :-)

Reply:
At this point we are 'opening' a node, turning one leaf node into 2 leaf nodes, hence the +1.